Mathlib/Util/CompileInductive.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2023 Parth Shastri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Parth Shastri
+-/
+import Lean.Compiler.CSimpAttr
+import Lean.Elab.Predefinition
+
+/-!
+# Define the `#compile inductive` command.
+-/
+
+namespace Mathlib.Util
+
+open Lean
+
+private def replaceConst (old new : Name) (e : Expr) : Expr :=
+  e.replace λ | .const n us => if n == old then some (.const new us) else none | _ => none
+
+open Meta
+
+private def mkFunExts' (xs : Array Expr) (e : Expr) (βfg : Expr × Expr × Expr) : MetaM Expr :=
+  Prod.fst <$> xs.foldrM (init := (e, βfg)) λ x (e, β, f, g) => do
+    let α ← inferType x
+    let f ← mkLambdaFVars #[x] f
+    let g ← mkLambdaFVars #[x] g
+    return (
+      mkApp5
+        (.const ``funext [(← inferType α).sortLevel!, (← inferType β).sortLevel!])
+        α
+        (← mkLambdaFVars #[x] β)
+        f
+        g
+        (← mkLambdaFVars #[x] e),
+      ← mkForallFVars #[x] β,
+      f,
+      g
+    )
+
+private def mkFunExts (e : Expr) : MetaM Expr := do
+  forallTelescope (← inferType e) λ xs body => do
+    let some βfg := (← whnf body).eq?
+      | throwError "expected equality"
+    mkFunExts' xs (mkAppN e xs) βfg
+
+private def mkEq (α a b : Expr) : MetaM Expr := do
+  return mkApp3 (.const ``Eq [(← inferType α).sortLevel!]) α a b
+
+private def mkEqRefl (α a : Expr) : MetaM Expr := do
+  return mkApp2 (.const ``Eq.refl [(← inferType α).sortLevel!]) α a
+
+open Elab
+
+/--
+Compile the definition `dv`.
```

Comment:
Could you explain what this does? Currently the docstring doesn't tell me anything I can't learn from the name and I actually don't know what this is for.